File: 056-merge-intervals/merge-intervals.py
# -*- coding:utf-8 -*-


# Given a collection of intervals, merge all overlapping intervals.
#
#
# For example,
# Given [1,3],[2,6],[8,10],[15,18],
# return [1,6],[8,10],[15,18].


# Definition for an interval.
# class Interval(object):
#     def __init__(self, s=0, e=0):
#         self.start = s
#         self.end = e

class Solution(object):
    def merge(self, intervals):
        """
        :type intervals: List[Interval]
        :rtype: List[Interval]
        """
        intervals.sort(key = lambda i: i.start)
        res = []
        for i in intervals:
            if res and i.start <= res[-1].end:
                res[-1].end = max(res[-1].end, i.end)
            else:
                res.append(i)
        return res
        

File: 056-merge-intervals/merge-intervals.js
// Given a collection of intervals, merge all overlapping intervals.
//
//
// For example,
// Given [1,3],[2,6],[8,10],[15,18],
// return [1,6],[8,10],[15,18].


/**
 * Definition for an interval.
 * function Interval(start, end) {
 *     this.start = start;
 *     this.end = end;
 * }
 */
/**
 * @param {Interval[]} intervals
 * @return {Interval[]}
 */
var merge = function(intervals) {
    if (!intervals.length)  return []
    
    intervals.sort((a, b) => a.start - b.start)
    const res = [intervals[0]]
    
    for (let i = 1; i < intervals.length; i += 1) {
        if (intervals[i].start <= res[res.length - 1].end) {
            res[res.length - 1].end = Math.max(intervals[i].end, res[res.length - 1].end)
        } else {
            res.push(intervals[i])
        }
    }
    
    return res
};
